### ISSUE TITLE
Provider logic does not correctly adhere to the specification

### ISSUE THREAD

**Issue by buehler** [nextauthjs/next-auth#4554]
### Environment

mentioned in https://github.com/nextauthjs/next-auth/issues/4543

### Reproduction URL

https://github.com/nextauthjs/next-auth/issues/4543

### Describe the issue

Based on https://github.com/nextauthjs/next-auth/issues/4543, next auth does not correctly act according to the specification. When using the provided configuration in the mentioned issue, the mentioned error occurs. This does not comply with the OIDC specification. It _does_ comply with the oauth2 authorization code flow (RFC6749), but not with the OIDC spec. Since OAuth has no userinfo endpoint, the OIDC spec defines that the returned id_token (which is required) contains the correct "sub" of a user. When then querying the user info endpoint for the profile, the sub MUST be verified. Otherwise, the content of the userinfo endpoint MUST NOT be used.

Currently, next auth throws an error.

As suggested in the issue by @balazsorban44: when using `idToken:false` and removing the `openid` scope, the whole provider and configuration does not adhere to the OIDC spec anymore. When using the suggested config, the following error is returned (by an IDP that adheres to the spec): 
```
[next-auth][error][OAUTH_CALLBACK_ERROR] 
https://next-auth.js.org/errors#oauth_callback_error invalid_scope {message: 'invalid_scope', stack: 'Error: invalid_scope
    at oAuthCallback (C:…dules\next\dist\server\base-server.js:477:37)', name: 'Error'}
```

### How to reproduce

described in https://github.com/nextauthjs/next-auth/issues/4543

### Expected behavior

Mentioned in https://github.com/nextauthjs/next-auth/issues/4543. I'd expect the behavior to be according to the specification. The lib must not throw an error when an ID token is present (when using the openid scope) and `idToken: false` is set.

I would like a constructive discussion about this topic because I think in #4543, there has been a misunderstanding.

This does not only diminish the usability and versatility of this library, but also introduces a security issue. There is an attack vector by token substitution when not verifying the sub in the id token.
(source: NOTE: Due to the possibility of token substitution attacks (see Section 16.11), the UserInfo Response is not guaranteed to be about the End-User identified by the sub (subject) element of the ID Token. The sub Claim in the UserInfo Response MUST be verified to exactly match the sub Claim in the ID Token; if they do not match, the UserInfo Response values MUST NOT be used.)

**Comment by buehler** [comment]
If this is not the correct way to report such an issue, please guide me where I can report this issue and have constructive discussion about it.

**Comment by GuiSim** [comment]
Intuit (and Quickbooks) use a JWK with RSA and 1024 bit modulus length.
You can see it here: https://oauth.platform.intuit.com/op/v1/jwks

Jose requires 2048 bit min for RSA: https://github.com/panva/jose/issues/210
The check is here: https://github.com/panva/jose/blob/main/src/runtime/node/check_modulus_length.ts#L68-L71

This means that when next-auth validates the JWT signature of the id_token, Jose throws and the validation fails. The modulus length is too short.
I have written to Intuit to ask them to increase the modulus length of their key, but I doubt this will get resolved quickly.

I thought that I could easily work around this problem by using `idToken:false`, ignoring the `id_token` and its signature check completely.
Sadly this isn't the case, as reported here setting `idToken` to `false` will result in an error if the IDP returns an `id_token`, which Intuit/Quickbooks does.

TL;DR: I could work around JWT security concerns if `idToken:false` was supported even with an `id_token` is found.

**Comment by GuiSim** [comment]
This was introduced by this commit: https://github.com/panva/node-openid-client/commit/92ffee5c63dc31fb578c731572cf0f83a7b53f1d

This is in `node-openid-client` so I'm unsure how much control `next-auth` has on this check, sadly.

**Comment by panva** [comment]
By setting `idToken: false` (which, granted, could be named better) a standard OAuth 2.0 flow without OpenID Connect is expected. Of course then requesting the `openid` scope and expecting pure OAuth 2.0 flow will throw because it now receives an unexpected ID Token.

By setting `idToken: true` an OIDC Authorization Code flow is used. Meaning that ID Token is required to be returned and all defined checks are present. Even [comparing](https://github.com/panva/node-openid-client/blob/v5.1.6/lib/client.js#L1293-L1299) the `ID Token` `sub` claim with the one returned by the userinfo endpoint.

All required checks are present, the exhibited client behaviours are conform to the OIDC specification. If you use an openid connect flow, then `idToken: true` must be used so that all those OpenID specific checks can be made. If that still fails, it's because the Identity Provider is not exhibiting conform behaviours, which is what I'm seeing here (use of 1024 bit RSA keys). That goes against the very core JSON Web Algorithm definition and will also be rejected for a very good reason.

Furthermore token substitution is not a vector here because Authorization Code flow is used and TLS is used during the code<>access token exchange which is also the only time the "userinfo" (which may be the OIDC userinfo_endpoint when idToken: true, but any arbitrary endpoint when idToken:false) is queried. There is place to substitute an access token anyway.

I understand your frustration but I am not seeing errors in the client-side behaviour.

The only issue I see are weak RSA 1024 bit keys which are correctly rejected.

**Comment by buehler** [comment]
Thanks for the clarification @panva. If I understand you correctly, using `idToken: true` in combination with an OIDC flow is the way to go. Do we need to call the user-info endpoint by ourselves? Because when the id token does not contain the user information, no information is fetched from the user info endpoint.

**Comment by panva** [comment]
That's something @balazsorban44 can probably speak to.

**Comment by balazsorban44** [comment]
I cannot see your code, but if the `id_token` does not contain the information you need, you could override the `userinfo` request handler as described in the docs: https://next-auth.js.org/configuration/providers/oauth#userinfo-option and make an extra request to whatever endpoint.

I agree that `idToken: boolean` did not make much sense here, and I plan to remove it, as detecting the presence of the `openid` scope should probably be enough now, and we are doing it already anyway. https://github.com/nextauthjs/next-auth/pull/3514

**Comment by dfenerski** [comment]
@GuiSim I am having the same problem with QuickBooks. Have you managed to find a way around it?

**Comment by GuiSim** [comment]
@dfenerski 
As a temporary hack I was able to get it to work by patching this https://github.com/panva/jose/blob/main/src/runtime/node/check_modulus_length.ts#L68-L71

with this:
```
    if (getModulusLength(key) < 1024) {
        throw new TypeError(`${alg} requires key modulusLength to be 1024 bits or larger`);
     }
```

Using https://www.npmjs.com/package/patch-package

I have not been able to get in touch with QuickBooks to get them to resolve this issue.
